tests/unit/pmodel/test_optimal_chi.py

```diff
@@ -0,0 +1,182 @@
+        ),
+        (
+            OptimalChiLavergne20C4,
+            dict(tc=20, patm=101325, co2=400, vpd=1000, theta=0.5),
+            dict(beta=24.97251, chi=0.44432, mc=0.28091, mj=1.0, mjoc=3.55989),
+        ),
+        (
+            OptimalChiC4NoGamma,
+            dict(tc=20, patm=101325, co2=400, vpd=1000),
+            dict(chi=0.3919, mc=0.25626, mj=1.0),
+        ),
+        (
+            OptimalChiC4NoGammaRootzoneStress,
+            dict(tc=20, patm=101325, co2=400, vpd=1000, rootzonestress=0.5),
+            dict(chi=0.31305, mc=0.21583, mj=1.0),
+        ),
+    ],
+)
+def test_subclasses(pmodelenv_args, subclass, expected):
+    """Test that subclasses work as expected."""
+    env = PModelEnvironment(**pmodelenv_args)
+    instance = subclass(env)
+    for key, value in expected.items():
+        assert getattr(instance, key) == pytest.approx(value, rel=1e-3)
+
+
+# @pytest.mark.parametrize(
+#     argnames=["subclass", "pmodelenv_args", "expected"],
+#     argvalues=[
+#         (
+#             OptimalChiPrentice14,
+#             dict(tc=np.nan, patm=101325, co2=400, vpd=1000),
+#             dict(chi=0.69435, mc=0.33408, mj=0.7123, mjoc=2.13211),
+#         ),
+#         (
+#             OptimalChiPrentice14RootzoneStress,
+#             dict(tc=20, patm=101325, co2=400, vpd=1000, rootzonestress=np.nan),
+#             dict(chi=0.62016),
+#         ),
+#         (
+#             OptimalChiC4,
+#             dict(tc=np.nan, patm=101325, co2=400, vpd=1000),
+#             dict(chi=0.44967, mj=1.0, mjoc=1.0),
+#         ),
+#         (
+#             OptimalChiC4RootzoneStress,
+#             dict(tc=20, patm=101325, co2=400, vpd=1000, rootzonestress=np.nan),
+#             dict(chi=0.37659, mj=1.0, mjoc=1.0),
+#         ),
+#         (
+#             OptimalChiLavergne20C3,
+#             dict(tc=20, patm=101325, co2=400, vpd=1000, theta=np.nan),
+#             dict(beta=224.75255, chi=0.73663, mc=0.34911, mj=0.7258, mjoc=2.07901),
+#         ),
+#         (
+#             OptimalChiLavergne20C4,
+#             dict(tc=20, patm=101325, co2=400, vpd=1000, theta=np.nan),
+#             dict(beta=24.97251, chi=0.44432, mc=0.28091, mj=1.0, mjoc=3.55989),
+#         ),
+#         (
+#             OptimalChiC4NoGamma,
+#             dict(tc=np.nan, patm=101325, co2=400, vpd=1000),
+#             dict(chi=0.3919, mc=0.25626, mj=1.0),
+#         ),
+#         (
+#             OptimalChiC4NoGammaRootzoneStress,
+#             dict(tc=20, patm=101325, co2=400, vpd=1000, rootzonestress=np.nan),
+#             dict(chi=0.31305, mc=0.21583, mj=1.0),
+#         ),
+#     ],
+# )
+# def test_nan_handling(pmodelenv_args, subclass, expected):
+#     """Test that subclasses handles NaNs correctly."""
+#     env = PModelEnvironment(**pmodelenv_args)
+#     instance = subclass(env)
+#     for key, value in expected.items():
+#         assert getattr(instance, key) == pytest.approx(value, rel=1e-3)
```

Comment:
We'd expect an assertion error here - the `np.nan` is propagated through the calculations so that the resulting expected values are also `np.nan`. It would be a serious problem if that didn't happen, so asserting that each attribute is `np.nan` is a good guard.

I'd actually be tempted to extend this and feed in `pmodelenv_args` as complete data and then have a loop that copies the dict and sets each value in turn to `np.nan`. That way we know that missing data in any input results in missing data. Something like:
```suggestion
def test_nan_handling(pmodelenv_args, subclass, expected):
    """Test that subclasses handles NaNs correctly."""
    for var in pmodelenv_args:
        pmodelenv_args_copy = pmodelenv_args.copy()
        pmodelenv_args_copy[var] = np.nan
        env = PModelEnvironment(**pmodelenv_args_copy)
        instance = subclass(env)
        for key, value in expected.items():
            assert getattr(instance, key) == pytest.approx(value, rel=1e-3)
```

Reply:
That last line should be `assert np.all(np.isnan(getattr(instance, key)))` or similar.

Reply:
Implemented.